AKS-Allow version upgrade only from current version to > 1 minor version

skanakal opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes  https://github.com/rancher/dashboard/issues/11884
<!-- Define findings related to the feature or bug issue. -->

### Occurred changes and/or fixed issues
<!-- Include information of the changes, including collateral areas which have been affected by this PR as requirement or for convenience. -->
filters the available versions to only include versions that are:
- The same minor version as originalVersion
- OR Upgrades to one version above

### Technical notes summary
<!-- Outline technical changes which may pass unobserved or may help to understand the process of solving the issue -->

### Areas or cases that should be tested
<!-- Areas that should be tested can include Airgap checks, Rancher upgrades, K8s upgrade, etc. -->
<!-- Which browser did you use for local testing? The reviewer should test with a different browser. -->
<!-- Add missing steps or rewrite them if have been missed or to complement existing information. This should define a clear way to reproduce it and not an approximation. -->
- Create AKS cluster from Rancher UI and select lowest supported k8s version
- Upgrade version to highest supported version

### Areas which could experience regressions
<!-- Create a detailed list of areas to be analyzed which may be affected by the changes, which would require a prior research to avoid regressions. -->

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->
![image](https://github.com/user-attachments/assets/6d37e823-4b74-4fa5-99f2-c378428109b1)


### Checklist
- [x] The PR is linked to an issue and the linked issue has a Milestone, or no issue is needed
- [x] The PR has a Milestone <!-- The milestone should automatically be assigned if the linked issue has one, but does not always happen (incorrectly linked, issue has no milestone, etc) -->
- [x] The PR template has been filled out
- [x] The PR has been self reviewed <!-- There are no TODOs, no incorrect files in the PR, all the required files are there, no commented out code, etc-->
- [x] The PR has a reviewer assigned
- [x] The PR has automated tests or clear instructions for manual tests and the linked issue has appropriate QA labels, or tests are not needed
- [x] The PR has reviewed with UX and tested in light and dark mode, or there are no UX changes

mantis-toboggan-md reviewed (changes requested):
I understand the thinking behind borrowing code from the EKS provisioning form, but EKS doesn't include patch versions and that code can't be straight copied here. Your screenshot shows that this doesn't work. Users should be able to upgrade from eg 1.31.x to 1.32.x

This would work:

<details>

```
aksVersionOptions(): { value: string, label: string, sort?: string, disabled?: boolean }[] {
      const validVersions = this.allAksVersions.reduce((versions, v: string) => {
        const coerced = semver.coerce(v);

        if (!coerced || (this.supportedVersionRange && !semver.satisfies(coerced.version, this.supportedVersionRange))) {
          return versions;
        }

        if (!this.originalVersion) {
          versions.push({ value: v, label: v });
        } else if (semver.lte(semver.coerce(this.originalVersion), coerced)) {
          const highestSupportedMinor = semver.coerce(this.originalVersion)?.minor + 1;

          if (highestSupportedMinor && highestSupportedMinor < coerced?.minor) {
            versions.push({
              value:    v,
              label:    `${ v } ${ this.t('aks.kubernetesVersion.upgradeWarning') }`,
              disabled: true
            });
          } else {
            const label = v === this.originalVersion ? this.t('aks.kubernetesVersion.current', { version: v }) : v;

            versions.push({ value: v, label });
          }
        }

        return versions;
      }, [] as { value: string, label: string, sort?: string, disabled?: boolean }[]);

      validVersions.forEach((v) => {
        v.sort = sortable(v.value);
      });

      const sorted = sortBy(validVersions, 'sort', true); // Descending order

      if (!this.config.kubernetesVersion) {
        const firstValid = sorted.find((v) => !v.disabled);

        this.config.kubernetesVersion = firstValid?.value;
      }

      return sorted;
    }

```

</details>
  
  There should also be a unit test added to pkg/aks/components/__tests__/Config.test.ts
  
<details>
    
```
  it.each([[
    '1.26.0',
    mockVersionsSorted.filter((v: string) => semver.lte(v, '1.27.0') && semver.gte(v, '1.26.0'))
  ],
  ['1.25.0',
    mockVersionsSorted.filter((v: string) => semver.lte(v, '1.26.0') && semver.gte(v, '1.25.0'))
  ],
  ])('should only allow upgrading one minor version at a time', async(originalVersion, validVersions) => {
    const wrapper = shallowMount(Config, {
      propsData: {
        config: DEFAULT_CLUSTER_CONFIG, value: {}, mode: _EDIT
      },
      ...requiredSetup({ value: '>=1.23.x' })
    });

    wrapper.setData({ originalVersion });

    await setCredential(wrapper);
    const versionDropdown = wrapper.getComponent('[data-testid="cruaks-kubernetesversion"]');
    const enabledOptions = versionDropdown.props().options.reduce((enabledOpts, opt) => {
      if (!opt.disabled) {
        enabledOpts.push(opt.value);
      }

      return enabledOpts;
    }, []);

    expect(enabledOptions).toStrictEqual(validVersions);
  });  
```
</details>

skanakal commented:
@mantis-toboggan-md Thank you for the feedback! I realize 🤦🏾‍♂️ I overlooked the point regarding the version upgrade logic. I’ve reviewed the changes now and implemented the updates to the aksVersionOptions() method to allow upgrades between minor versions... Additionally, I’ve added the requested unit test to verify that only one minor version upgrade is allowed at a time.

mantis-toboggan-md reviewed (approved):
LGTM thanks for the quick response @skanakal. I'll merge this PR for you once e2e tests pass

mantis-toboggan-md commented:
Apologies @skanakal but could you rebase your PR? We had a bug in an e2e test and this PR will need to bring in the now-merged fix to pass ci gates
